rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,442 @@
+
+### Sync Commit
+
+In contrast to the read side which may mix the positions in term of "durability" and "commit", the write / commit side is more straightforward, it only need to track the things about durability.
+
+To implement Synchronous Commit semantics, the write side remains do not need to change at all: the operation returns once the WAL is persisted according to the specified durability level. A `DurabilityLevel` is all we need here.
+
+Like:
+
+```rust
+let opts = WriteOptions::new().with_await_durability(DurabilityLevel::Remote);
+db.write(key, value, opts).await?;
+```
+
+In other systems, this option is often named as `sync`, to emphasize the idea of synchronous commit. To respect the idea of Least Surprise, it might be not a bad idea to rename `await_durability` to `sync`.
+
+```rust
+let opts = WriteOptions::new().with_sync(DurabilityLevel::Memory);
+db.write(key, value, opts).await?;
+```
+
+However, for the writes that not expected to durablely persisted, this can be thought of as **disabling** synchronous commit entirely rather than synchronizing to memory. The code sample may be like:
+
+```rust
+enum SyncLevel {
+    Off,
+    Local,
+    Remote,
+}
+
+let opts = WriteOptions::new().with_sync(SyncLevel::Off);
+db.write(key, value, opts).await?;
+```
+
+The benefits of this naming change are:
+
+1. The name "Synchronous Commit" directly describes the feature we aim to support.
+2. The term "sync" is commonly used across database systems and is well understood, familiar to users.
+3. The names are shorter, easier to type & read.
+4. It's accurate to describe as sync commit as "off" when it comes with memory durability.
+
+The inner implementation of the write side do not need to change, it's still just await the data to be persisted to storage as the specified durability level, then return. All we put into discussion here is the naming stuff.
+
+### Difference between "await_durable" and "sync"
+
+`await_durable` and `sync` appear to have similar behaviors - on the write side, both wait until data reaches a persistent state before returning.
+
+However, there are subtle semantic differences between them:
+
+1. `sync` determines when data enters the committed state and returns after data is committed.
+2. `await_durable` only waits for the data's durability state.
+
+For example, when a write has `Sync::Off` set, it is considered to enter the Committed state immediately, this write will become visible to readers with `LastCommitted` immediately.
+
+However, at some point this data will still be eventually flushed to storage if the DB keeps running. If you want to subscribe or wait for when this `sync:Off` write becomes durably persisted, you can still use `await_durable` to achieve this goal.
+
+Also, when using `sync` commit, it's possible to queue the later Non-Sync writes to be queued in the WAL buffer, delaying them later to be applied to MemTable and visible to readers.
```

Comment:
I don't quite understand this comment. Can you explain it a bit?

Reply:
consider we have the following sequences:

- op 1: write key1=val1, with sync=Remote
- op 2: write key2=val2, with sync=Off

it's ordered to apply these operations to the memtable, which means to be visible to other readers.

the op 2 has to await the op1 to be synced to remote before the write from op 2 become visible to users, even it disables sync commit with `sync=Off`